thunder/tests/opinfos.py

```diff
@@ -4345,6 +4345,57 @@ def take_along_axis_sample_generator(op, device, dtype, requires_grad, **kwargs)
 shape_ops.append(take_along_axis_opinfo)
 
 
+def gather_sample_generator(op, device, dtype, requires_grad, **kwargs):
+    make = partial(make_tensor, device=device, dtype=dtype, requires_grad=requires_grad)
+    # torch.gather expects index to be long but not int
+    # Index is not differentiable! Marking requires_grad as False
+    make_index = partial(make_tensor, device=device, dtype=torch.long, requires_grad=False)
+
+    for shape_a, dim, shape_b in take_along_axis_cases:
+        canonicalized_dim = dim if dim >= 0 else dim + len(shape_a)
+        a = make(shape_a)
+        b = make_index(shape_b, low=0, high=shape_a[dim])
+        yield SampleInput(a, index=b, dim=dim)
+
+    # Questionable use case. Do we want to support these?!
+    # Note that gather doesn't have the broadcast requirement, it only requires
+    # 1. a.shape[i]      >= index.shape[i] for i != dim
+    #
+    # a.shape, dim, index.shape
+    scatter_add_cases = (
+        ((4, 5, 3), 0, (3, 2, 3)),
+        ((4, 5, 3), 1, (3, 5, 2)),
+        ((4, 5, 3), 2, (3, 2, 8)),
+    )
+    for shape_a, dim, shape_b in scatter_add_cases:
+        a = make(shape_a)
+        b = make_index(shape_b, low=0, high=shape_a[dim])
+        yield SampleInput(a, index=b, dim=dim)
+
+
+gather_opinfo = OpInfo(
+    ltorch.gather,
+    supports_grad=True,
+    sample_input_generator=gather_sample_generator,
+    torch_reference=torch.gather,
+    test_directives=(
+        # Torch doesn't support complex half on gather
+        DecorateInfo(
+            pytest.mark.skip,
+            "test_core_vs_torch_consistency",
+            dtypes=(datatypes.complex32,),
+        ),
+        DecorateInfo(
+            custom_comparator(partial(assert_close, atol=1e-1, rtol=1e-1)),
```

Comment:
The tolerances are a bit too high it seems, what's the failure rate with something smaller?

Reply:
Yeah, this is strange. Gather just copies, so if this tolerance is needed it is likely that something isn't right..

Reply:
Ah, well, I guess the issue is backward? Well, the indices gotta be unique. Only then the impl backward is correct.

Reply:
I remove the test_directives; The tests pass.

Reply:
@rdspring1 , it is still weird, as the backward formula assumes no repeated indices. Otherwise it is incorrect. Do we create such index tensors by chance? I am not sure we fix the seed just yet, so, maybe, it is best to resolve that right now?